### ISSUE TITLE
Can't get Featured image

### ISSUE THREAD

**Issue by ChristianGiupponi** [jgrossi/corcel#129]
I have a custom post type called "dicono-di-noi" and if I try to get the entries with:

    $dicono_di_noi = DiconoDiNoi::published()->with('attachment')->orderBy(DB::raw('RAND()'))->get();

I'm not able to get the image.

This is my view:

```
 @foreach($dicono_di_noi as $dicono)
        <div class="item">
            {!! $dicono->image !!}
        </div>
  @endforeach
```

I used `$dicono->image` according to this issue: https://github.com/jgrossi/corcel/issues/94 but this is the result:

~~~
SQLSTATE[42S02]: Base table or view not found: 1146 Table 'testsite.postmeta' doesn't exist (SQL: select * from `postmeta` where `post_id` is null and `post_id` is not null)
~~~

**Comment by jgrossi** [comment]
Hi @ChristianGiupponi are you working with a custom connection name called `testsite`, right? Is this set in your custom post type class?

**Comment by ChristianGiupponi** [comment]
Hi @jgrossi the connection is named `wordpress`:

~~~
'wordpress' => [
            'driver'    => 'mysql',
            'host'      => env('DB_HOST', 'localhost'),
            'port'      => env('DB_PORT', '3306'),
            'database'  => env('DB_DATABASE', 'forge'),
            'username'  => env('DB_USERNAME', 'forge'),
            'password'  => env('DB_PASSWORD', ''),
            'charset'   => 'utf8',
            'collation' => 'utf8_unicode_ci',
            'prefix'    => 'wp_',
            'strict'    => false,
            'engine'    => null,
        ],
~~~

and this is my model:

~~~
<?php

namespace App\Models\Wordpress;

use Caffeinated\Presenter\Traits\PresentableTrait;
use Corcel\Attachment;
use Corcel\Post as Corcel;

class Testimonial extends Corcel
{
    use PresentableTrait;

    /**
     * @var string
     */
    protected $presenter = 'App\Presenters\Testimonials\Testimonial';

    /**
     * @var string
     */
    protected $postType = 'testimonial';

    /**
     * @var string
     */
    protected $connection = 'wordpress';

}
~~~

right now to get the image this is the only thing that works for me:

~~~
foreach($testimonial->attachment as $attachment){
            $image = $attachment->guid;
            $name = $attachment->post_name;
        }
~~~

**Comment by jgrossi** [comment]
@ChristianGiupponi here is the code to get the image. Please check if the `_thumbnail_id` meta exists for this post:

```
/**
     * Gets the featured image if any
     * Looks in meta the _thumbnail_id field.
     *
     * @return string
     */
    public function getImageAttribute()
    {
        if (!is_null($this->meta->_thumbnail_id)) {
            $image = Attachment::find($this->meta->_thumbnail_id);

            if (!is_null($image)) {
                return $image->guid;
            }
        }
    }
```

Thanks.

**Comment by ChristianGiupponi** [comment]
Should I add this to my Model right?

**Comment by jgrossi** [comment]
No @ChristianGiupponi this method looks for a `image` attribute, not `imageAttribute`. We missed the tests for this feature I think.

The error you're getting is about the database `testsite` and the table `postmeta`. The weird here is you've set the connection to `wordpress` and it has the `wp_` preffix, and the class is looking for `postmeta`, not `wp_postmeta`.

**Comment by ChristianGiupponi** [comment]
@jgrossi yeah, sorry, I forgot to remove the `attribute`.  
I have tried again but nothing changed, for some reason it ignore the `wp_` and set all to `null`

~~~
SQLSTATE[42S02]: Base table or view not found: 1146 Table 'testsite.postmeta' doesn't exist (SQL: select * from `postmeta` where `post_id` is null and `post_id` is not null)
~~~

**Comment by jgrossi** [comment]
Is the database name of your default connection the same `testsite`?

I think I know what's happening. When using `->attachment` we are using the `hasMany()` method from `Corcel\Model` so it forwards the `connection` name to all the relationships, but when using `->image` it looks this is not working. This method looks for `$this->meta->_thumbnail_id` and get the corresponding attachment. 

Can you check if `$this->meta->_thumbnail_id` is an integer?

**Comment by ChristianGiupponi** [comment]
The default connection is `mysql`, the laravel default one and yes, the `$this->meta->_thumbnail_id` is an integer.


I've investigate in `Attachment` and I've noticed that `Corcel\Post` to not have the `$connection` properties.

I have added 

~~~
protected $connection = 'wordpress';
~~~

and now the code print the image url, could be this the problem?

**Comment by jgrossi** [comment]
No, `Corcel` should not have this attribute, because this must be set by your `Testimonial` class, and it extends `Corcel\Post`. So this is really a problem of connection config of Corcel.

You got the image url using `->image` or `->attachment`?

**Comment by ChristianGiupponi** [comment]
If I remove the connection from `Corcel\Post` it stop working.

I get it with `->image`.

BTW: I'm using `dev-master`, could this be the problem?

**Comment by jgrossi** [comment]
I don't think so. The `master` branch is updated with the last changes we've made. 

Please, can you debug `Corcel\Post::getImageAttribute()`? You've said the `$this->meta->_thumbnail_id` is an integer, so the problem should be in `Attachment::find()` call. I think when calling this is where the error is being triggered.

Can you check that?

**Comment by ChristianGiupponi** [comment]
Yes, the error is here, I have tried to add the number directly and it still fail.  
What should I check?

**Comment by jgrossi** [comment]
I think `Attachment` does not know the connection, but that's weird because you said you set the `$connection` to `Corcel\Post` and it works.

**Comment by ChristianGiupponi** [comment]
Yes, but in my code I do not extends Attachment anywhere...should I extent it? Maybe is for this reason that do not see the `$connection`.

**Comment by jgrossi** [comment]
You should not extend `Attachment`, you only should have to extend your custom post type to `extends Corcel\Post` and just it.

So when you set `$connection` in `Corcel\Post` were you extending `Attachment`? Is this correct?

**Comment by ChristianGiupponi** [comment]
When I set `$connection` in `Corcel\Post` I'm extending `Model`:

`class Post extends Model`.

and

`Attachmend extends Post`

**Comment by jgrossi** [comment]
Try to make `Attachment` extending `Model` instead of `Post` and check if it'll work please.

**Comment by ChristianGiupponi** [comment]
If I extend `Model` I get:

~~~
SQLSTATE[42S02]: Base table or view not found: 1146 Table 'testsite.attachments' doesn't exist (SQL: select * from `attachments` where `attachments`.`id` = 53 limit 1)
~~~

If you want to test I sued this plugin for testimonials: https://wordpress.org/plugins/easy-testimonials/

**Comment by jgrossi** [comment]
Yep, that's definitely a connection problem. Sorry, but we have to check that first, maybe using `getImageAttribute()` should not be the best solution. ASAP I'm going to check that, but if you want to, any help is welcome.

**Comment by rundef** [comment]
Looks like the connection isn't passed to the [Attachment model](https://github.com/jgrossi/corcel/blob/master/src/Post.php#L403).

Maybe something like this could fix that issue ?

```php
$attachment = new Attachment();
$attachment->setConnection($this->getConnection()->getName());
$image = $attachment->find($this->meta->_thumbnail_id);
```

**Comment by jgrossi** [comment]
@ChristianGiupponi can you  make the test @rundef said? I cannot do that at the moment.

Replace `$image = Attachment::find($this->meta->_thumbnail_id);` line 403 in `Corcel\Post` to:

```
$attachment = new Attachment();
$attachment->setConnection($this->getConnection()->getName());
$image = $attachment->find($this->meta->_thumbnail_id);
```

Thanks!

**Comment by jgrossi** [comment]
And thanks @rundef

**Comment by ChristianGiupponi** [comment]
@rundef this fix the issue!

**Comment by jgrossi** [comment]
Awesome @ChristianGiupponi and thanks a lot @rundef. Can anyone send the PR? I can do that only in 1 or 2 hours, so if you can, just send. Thanks again

**Comment by ChristianGiupponi** [comment]
Sure, just 2 minutes and I send it

**Comment by jgrossi** [comment]
Thanks @ChristianGiupponi

**Comment by jgrossi** [comment]
Fixed in `v1.0.5` ;-)

**Comment by ChristianGiupponi** [comment]
Now the problem is the huge amount of query for each image.  
I have tried to implement this: https://github.com/jgrossi/corcel/issues/107 but seems that it do not use the right connection:

~~~
SQLSTATE[42S02]: Base table or view not found: 1146 Table 'testsite.postmeta' doesn't exist (SQL: select * from `postmeta` where `meta_key` = _thumbnail_id and `postmeta`.`post_id` in (56, 52))
~~~
